spec/postgres_runner_helper.rb

```diff
@@ -0,0 +1,142 @@
+require "etc"
+require "tmpdir"
+require "socket"
+require "fileutils"
+require "singleton"
+
+require_relative 'support/mixins/pg_environment_updater'
+
+# This class is in charge of booting additional instance of postgres that can
+# be used for validating features of PostgresAdmin.
+#
+# It makes use of `pg_ctl` to init, start and stop the instance, in is built so
+# that it can be used as a stub for LinuxAdmin as well.  While this is a
+# singleton, this really should only booted at the beginning of the suite, and
+# shutdown and cleaned up prior to exiting.
+class PostgresRunner
+  include Singleton
+  include FileUtils
+  include PgEnvironmentUpdater
+
+  attr_accessor :db_dir, :port
+
+  def self.start
+    instance.start
+  end
+
+  def self.stop
+    instance.stop
+  end
+
+  def self.running?
+    instance.running?
+  end
+
+  def self.port
+    instance.port
+  end
+
+  def self.set_env
+    instance.set_env
+  end
+
+  def self.destroy
+    instance.stop
+    instance.clean
+  end
+
+  def self.user
+    Etc.getlogin
+  end
+
+  def self.group
+    Etc.getgrgid(Etc.getpwnam(Etc.getlogin).gid).name
+  end
+
+  def initialize
+    initdb
+    set_env
+    set_port
+    start
+    create_root_role
+    create_postgres_role
+  end
+
+  def start
+    return if running?
+    fix_pg_conf
+    system("#{pg_ctl} start -D #{db_dir} -wo '-p #{port}'", :out => File::NULL)
+  end
+
+  def stop
+    return unless running?
+    system("#{pg_ctl} stop -D #{db_dir} -wm fast", :out => File::NULL)
+  end
+
+  def clean
+    remove_entry(db_dir)
+  end
+
+  def running?
+    return false unless db_dir
+    # avoiding a shellout by just checking the existance of the pg_ctl pid file
+    #
+    # see `man pg_ctl` in the FILES section for the description of this file
+    File.exist?(pid_file)
+  end
+
+  def set_env
+    @user = self.class.user # used by PgEnvironmentUpdater
+    ENV["APPLIANCE_PG_DATA"]    = db_dir
+    ENV["APPLIANCE_PG_SERVICE"] = "local_pg_instance"
+  end
+
+  private
+
+  def set_port
+    @port = TCPServer.open("127.0.0.1", 0) { |sock| sock.addr[1] }
```

Comment:
Do we really have to worry about ports? I would figure that we would be connecting to the local socket rather than over the network.

Reply:
On dev, there would be a port conflict if we left it to the defaults (I am pretty sure), and `postgres` wouldn't boot.